ff/src/fields/models/fp/mod.rs

```diff
@@ -0,0 +1,1015 @@
+use ark_serialize::{
+    buffer_byte_size, CanonicalDeserialize, CanonicalDeserializeWithFlags, CanonicalSerialize,
+    CanonicalSerializeWithFlags, EmptyFlags, Flags, SerializationError,
+};
+use ark_std::{
+    cmp::{Ord, Ordering, PartialOrd},
+    fmt::{Display, Formatter, Result as FmtResult},
+    io::{Read, Result as IoResult, Write},
+    marker::PhantomData,
+    ops::{Add, AddAssign, Div, DivAssign, Mul, MulAssign, Neg, Sub, SubAssign},
+    str::FromStr,
+    One, Zero,
+};
+
+#[macro_use]
+mod montgomery_backend;
+pub use montgomery_backend::*;
+
+use crate::{
+    BigInt, BigInteger, FftField, Field, FromBytes, LegendreSymbol, PrimeField, SquareRootField,
+    ToBytes,
+};
+/// A trait that specifies the configuration of a prime field.
+/// Also specifies how to perform arithmetic on field elements.
+pub trait FpConfig<const N: usize>: crate::FftConfig<Field = Fp<Self, N>> {
+    /// The modulus of the field.
+    const MODULUS: crate::BigInt<N>;
+
+    /// A multiplicative generator of the field.
+    /// `Self::GENERATOR` is an element having multiplicative order
+    /// `Self::MODULUS - 1`.
+    const GENERATOR: Fp<Self, N>;
+
+    /// Additive identity of the field, i.e. the element `e`
+    /// such that, for all elements `f` of the field, `e + f = f`.
+    const ZERO: Fp<Self, N>;
+
+    /// Multiplicative identity of the field, i.e. the element `e`
+    /// such that, for all elements `f` of the field, `e * f = f`.
+    const ONE: Fp<Self, N>;
+
+    /// Set a += b.
+    fn add_assign(a: &mut Fp<Self, N>, b: &Fp<Self, N>);
+
+    /// Set a -= b.
+    fn sub_assign(a: &mut Fp<Self, N>, b: &Fp<Self, N>);
+
+    /// Set a = a + a.
+    fn double_in_place(a: &mut Fp<Self, N>);
+
+    /// Set a *= b.
+    fn mul_assign(a: &mut Fp<Self, N>, b: &Fp<Self, N>);
+
+    /// Set a *= b.
+    fn square_in_place(a: &mut Fp<Self, N>);
+
+    /// Compute a^{-1} if `a` is not zero.
+    fn inverse(a: &Fp<Self, N>) -> Option<Fp<Self, N>>;
+
+    /// Compute the square root of a, if it exists.
+    fn square_root(a: &Fp<Self, N>) -> Option<Fp<Self, N>> {
+        // https://eprint.iacr.org/2012/685.pdf (page 12, algorithm 5)
+        // Actually this is just normal Tonelli-Shanks; since `P::Generator`
```

Comment:
```suggestion
        // Actually this is just normal Tonelli-Shanks; since `P::GENERATOR`
```

Should we keep the `P`? It is originally for the parameters.